File: 3259.超级饮料的最大强化能量.py
#
# @lc app=leetcode.cn id=3259 lang=python3
#
# [3259] 超级饮料的最大强化能量
#

# @lc code=start
class Solution:
    def maxEnergyBoost(self, energyDrinkA: List[int], energyDrinkB: List[int]) -> int:
        n = len(energyDrinkA)
        energyDrink = (energyDrinkA, energyDrinkB)

        @cache  # 缓存装饰器，避免重复计算 dfs 的结果（记忆化）
        def dfs(i: int, j: int) -> int:
            if i < 0:
                return 0
            res1 = dfs(i - 1, j) + energyDrink[j][i]
            res2 = dfs(i - 2, j ^ 1) + energyDrink[j][i]
            return max(res1, res2)
        
        return max(dfs(n - 1, 0), dfs(n - 1, 1))
# @lc code=end

File: 3259.超级饮料的最大强化能量.cpp
/*
 * @lc app=leetcode.cn id=3259 lang=cpp
 *
 * [3259] 超级饮料的最大强化能量
 */

// @lc code=start
class Solution
{
public:
    long long maxEnergyBoost(vector<int> &energyDrinkA, vector<int> &energyDrinkB)
    {
        int n = energyDrinkA.size();
        vector<array<long long, 2>> dp(n + 2);
        // 状态转移
        for (int i = 0; i < n; i++)
        {
            dp[i + 2][0] = max(dp[i + 1][0], dp[i][1]) + energyDrinkA[i];
            dp[i + 2][1] = max(dp[i + 1][1], dp[i][0]) + energyDrinkB[i];
        }
        return max(dp[n + 1][0], dp[n + 1][1]);
    }
};
// @lc code=end
